### PR TITLE
James changes

### PR DESCRIPTION
**Pull request by jminuse**

### CODE REVIEW

File: khan/model/nn.py [hunk 1]
```diff
@@ -55,15 +56,51 @@ def __init__(self, features, layer_sizes, atom_type="", prefix=""):
...
                 #A = tf.multiply( tf.nn.leaky_relu(A, alpha=0.2), tf.truncated_normal(shape=[y], mean=1.0, stddev=0.01) )
+                # CELU activation
+                posA = tf.cast(tf.greater_equal(A,0),tf.float32) * A
+                negA = tf.cast(tf.less(A,0),tf.float32) * A
+                A = posA + self.alpha * ( tf.exp(negA/self.alpha) - 1 )
 
             self.Ws.append(W)
             self.bs.append(b)
             self.As.append(A)
+        '''
```
Comment (proteneer) [inline khan/model/nn.py]: do you want this still?

File: khan/model/nn.py [hunk 2]
```diff
@@ -55,15 +56,51 @@ def __init__(self, features, layer_sizes, atom_type="", prefix=""):
...
+                #A = A * tf.nn.sigmoid(A) # "swish" activation
+                #A += (tf.sqrt(A**2+1) - 1)*0.5 # "bent identity" activation function (like smooth leaky relu - unbounded, unlike ELU)
+                #A = tf.exp(-self.freq * A**2) # Gaussian activation
+                #A = tf.sin(A)
+                #A = tf.nn.elu(A) # ELU, kind of like RELU but smooth
+                #if idx > 1: A = tf.nn.dropout(A, 0.9) # dropout
                 #A = tf.add( tf.nn.leaky_relu(A, alpha=0.2), tf.truncated_normal(shape=[y], stddev=0.001) ) # noisy leaky RELU
                 #A = tf.multiply( tf.nn.leaky_relu(A, alpha=0.2), tf.truncated_normal(shape=[y], mean=1.0, stddev=0.01) )
+                # CELU activation
+                posA = tf.cast(tf.greater_equal(A,0),tf.float32) * A
```
Comment (proteneer) [inline khan/model/nn.py]: ugh this is a little gnarly. I don't think I need to cast this to tf.float64 in the case of double precision training, and it's a damn shame there's no tf.nn.celu

File: khan/model/nn.py [hunk 3]
```diff
@@ -33,19 +33,20 @@ def __init__(self, features, layer_sizes, atom_type="", prefix=""):
         self.As = [features] # neurons
         self.atom_type = atom_type
 
+        self.alpha = tf.get_variable(prefix+'_'+atom_type+'_alpha', tuple(), tf.float32, tf.constant_initializer(0.1), trainable=False)
```
Comment (proteneer) [inline khan/model/nn.py]: is there a reason why this is a tf variable as opposed to a constant? Do you changing this value during the runs?

File: khan/training/trainer_multi_tower.py [hunk 4]
```diff
@@ -129,6 +132,7 @@ def __init__(self,
         towers,
         layer_sizes=(128, 128, 64, 8, 1),
         fit_charges=False,
+        charge_layer_sizes=(128, 128, 64, 8, 1),
```
Comment (proteneer) [inline khan/training/trainer_multi_tower.py]: charge fitting is currently broken so i'm not merging in charge changes with this round.

File: khan/training/trainer_multi_tower.py [hunk 5]
```diff
@@ -175,16 +179,16 @@ def __init__(self,
         self.sess = sess
 
         with tf.device('/cpu:0'):
-
-            self.learning_rate = tf.get_variable('learning_rate', tuple(), tf.float32, tf.constant_initializer(1e-3), trainable=False)
-            self.optimizer = tf.train.AdamOptimizer(
+            self.learning_rate = tf.get_variable('learning_rate', tuple(), tf.float32, tf.constant_initializer(1e-4), trainable=False)
+            self.optimizer = NadamOptimizer( # tf.train.AdamOptimizer(
```
Comment (proteneer) [inline khan/training/trainer_multi_tower.py]: Should we always use Nadam from now on?

File: khan/training/trainer_multi_tower.py [hunk 6]
```diff
@@ -319,7 +325,7 @@ def __init__(self,
         ws = self._weight_matrices()
         max_norm_ops = []
         for w in ws:
-            max_norm_ops.append(tf.assign(w, tf.clip_by_norm(w, 3.0, axes=1)))
+            max_norm_ops.append(tf.assign(w, tf.clip_by_norm(w, 2.0, axes=1)))
```
Comment (proteneer) [inline khan/training/trainer_multi_tower.py]: Do we still use max norms? If not I'll remove this entirely.

File: khan/training/trainer_multi_tower.py [hunk 7]
```diff
@@ -563,12 +569,11 @@ def submitter():
             # gpu   0 1 2
             # bid0  1 1 1
             # bid1  1 1 0
-
             try:
 
                 n_batches = dataset.num_batches(batch_size)
 
-                for b_idx, (mol_xs, mol_idxs, mol_yts) in enumerate(dataset.iterate(batch_size=batch_size, shuffle=shuffle)):
+                for b_idx, (mol_xs, mol_idxs, mol_yts) in enumerate(dataset.iterate(batch_size=batch_size, shuffle=shuffle, fuzz=1e-5)):
```
Comment (proteneer) [inline khan/training/trainer_multi_tower.py]: Does fuzz depend on precision? If we're training in double we could probably drop down to 1e-6 or more

Comment (proteneer) [reply khan/training/trainer_multi_tower.py]: note that the error in single precision is around 1e-4, so tuning this in the 1e-5 range is probably just create complete noise

File: khan/training/trainer_multi_tower.py [hunk 8]
```diff
@@ -609,3 +615,62 @@ def submitter():
 
         for i in range(n_tower_batches):
             yield self.sess.run(target_ops)
+
+    # run the actual training
+    def train(self, save_dir, rd_train, rd_test, rd_gdb11, eval_names, eval_datasets, eval_groups, batch_size, max_local_epoch_count=25, max_batch_size=1e4, max_global_epoch_count=1000):
```
Comment (proteneer) [inline khan/training/trainer_multi_tower.py]: Have you finalized this part of the code to be "canonical way" to train?

File: main_2.py [hunk 9]
```diff
@@ -24,8 +22,8 @@ def get_available_gpus():
     return [x.name for x in local_device_protos if x.device_type == 'GPU']
 
 def main():
-    avail_gpus = get_available_gpus()
-    print("Available GPUs:", avail_gpus)
+    #avail_gpus = get_available_gpus()
```
Comment (proteneer) [inline main_2.py]: I'm going to merge this in as-is without a CR since I'm not maintaining consumer space code

File: khan/model/nn.py (hunk 2 above)
Comment (jminuse) [reply khan/model/nn.py]: We could request tf.nn.celu, or put it in contrib. It's a tiny code change from ELU.

File: khan/model/nn.py (hunk 3 above)
Comment (jminuse) [reply khan/model/nn.py]: I have made it trainable for some test runs. It doesn't really need to be tf.variable in general.

File: khan/training/trainer_multi_tower.py (hunk 5 above)
Comment (jminuse) [reply khan/training/trainer_multi_tower.py]: I recommend NAdam, yes. Nesterov momentum is strictly better than regular momentum, people seem to think now.

File: khan/training/trainer_multi_tower.py (hunk 6 above)
Comment (jminuse) [reply khan/training/trainer_multi_tower.py]: I still use max norms! Without that, we have no regularization at all for points far from our dataset (points near the dataset are covered by input fuzzing).

File: khan/training/trainer_multi_tower.py (hunk 7 above)
Comment (jminuse) [reply khan/training/trainer_multi_tower.py]: No, the figure 1e-5 is determined by typical atomic forces. It's not related to the precision. And this is on the scale of the inputs, not the outputs, so the precision might not behave like you expect.

File: khan/training/trainer_multi_tower.py (hunk 8 above)
Comment (jminuse) [reply khan/training/trainer_multi_tower.py]: I think so. We'll see what Leif gets.

**Comment by proteneer** [comment]
Closing this via #70

**Pull request closed without merging**